File: riveredge-frontend/src/apps/kuaiplm/services/material-review.ts
/**
 * 物料评审 API（R-15 #37）
 */

import { api } from '../../../services/api';

const BASE = '/apps/kuaiplm/material-reviews';

export type MaterialReviewStatus = 'draft' | 'pending' | 'approved' | 'rejected';
export type MaterialUsageStatus = 'preferred' | 'limited' | 'forbidden';

export interface MaterialReviewLine {
  id?: number;
  uuid?: string;
  line_no?: number;
  material_id?: number | null;
  material_code: string;
  material_name: string;
  usage_status: MaterialUsageStatus | string;
  remarks?: string | null;
}

export interface MaterialReview {
  id: number;
  uuid: string;
  review_code: string;
  project_id: number;
  project_code: string;
  project_name: string;
  title: string;
  status: MaterialReviewStatus;
  remarks?: string | null;
  line_count?: number;
  lines?: MaterialReviewLine[];
  submitted_at?: string | null;
  approved_at?: string | null;
  created_at?: string;
  updated_at?: string;
  created_by_name?: string | null;
  updated_by_name?: string | null;
}

export interface MaterialReviewPayload {
  project_id: number;
  title: string;
  review_code?: string;
  remarks?: string | null;
  lines?: MaterialReviewLine[];
}

function unwrapList(res: unknown): { items: MaterialReview[]; total: number } {
  const r = (res || {}) as Record<string, unknown>;
  const items = (r.items ?? r.data ?? []) as MaterialReview[];
  return { items: Array.isArray(items) ? items : [], total: Number(r.total ?? items.length) };
}

export const materialReviewApi = {
  list: async (params: {
    skip?: number;
    limit?: number;
    keyword?: string;
    status?: string;
    project_id?: number;
  }) => {
    const res = await api.get(BASE, { params });
    return unwrapList(res);
  },
  get: async (id: number) => {
    const res = await api.get(`${BASE}/${id}`);
    return res as MaterialReview;
  },
  create: async (data: MaterialReviewPayload) => {
    const res = await api.post(BASE, data);
    return res as MaterialReview;
  },
  update: async (id: number, data: Partial<MaterialReviewPayload>) => {
    const res = await api.put(`${BASE}/${id}`, data);
    return res as MaterialReview;
  },
  submit: async (id: number) => {
    const res = await api.post(`${BASE}/${id}/submit`);
    return res as MaterialReview;
  },
  approve: async (id: number) => {
    const res = await api.post(`${BASE}/${id}/approve`);
    return res as MaterialReview;
  },
  reject: async (id: number) => {
    const res = await api.post(`${BASE}/${id}/reject`);
    return res as MaterialReview;
  },
  remove: async (id: number) => {
    await api.delete(`${BASE}/${id}`);
  },
};

File: riveredge-backend/src/apps/kuaiplm/schemas/material_review.py
"""物料评审 Schema（R-15 #37）"""

from __future__ import annotations

from datetime import datetime
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class MaterialReviewLineIn(BaseModel):
    material_id: Optional[int] = None
    material_code: str = Field(..., min_length=1, max_length=80)
    material_name: str = Field(..., min_length=1, max_length=200)
    usage_status: str = Field(..., description="preferred/limited/forbidden")
    remarks: Optional[str] = Field(None, max_length=500)


class MaterialReviewLineOut(MaterialReviewLineIn):
    model_config = ConfigDict(from_attributes=True)

    id: int
    uuid: UUID
    line_no: int


class MaterialReviewCreate(BaseModel):
    project_id: int
    title: str = Field(..., min_length=1, max_length=200)
    review_code: Optional[str] = Field(None, max_length=50)
    remarks: Optional[str] = None
    lines: List[MaterialReviewLineIn] = Field(default_factory=list)


class MaterialReviewUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    remarks: Optional[str] = None
    lines: Optional[List[MaterialReviewLineIn]] = None


class MaterialReviewResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    uuid: UUID
    review_code: str
    project_id: int
    project_code: str
    project_name: str
    title: str
    status: str
    remarks: Optional[str] = None
    submitted_at: Optional[datetime] = None
    approved_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    created_by: Optional[int] = None
    created_by_name: Optional[str] = None
    updated_by: Optional[int] = None
    updated_by_name: Optional[str] = None
    lines: List[MaterialReviewLineOut] = Field(default_factory=list)


class MaterialReviewListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    uuid: UUID
    review_code: str
    project_id: int
    project_code: str
    project_name: str
    title: str
    status: str
    line_count: int = 0
    created_at: datetime
    updated_at: datetime
    created_by_name: Optional[str] = None
    updated_by_name: Optional[str] = None


class MaterialReviewListResponse(BaseModel):
    items: List[MaterialReviewListItem]
    total: int
